File: app/lib/shared/services/habit_service.dart
import 'package:dio/dio.dart';

import '../../core/constants/api_constants.dart';
import '../../data/models/event_model.dart';
import '../../data/models/habit_model.dart';
import 'user_scope.dart';

/// HTTP client for Reclaim-style habit CRUD and scheduling.
class HabitService {
  HabitService({Dio? dio}) : _dio = dio ?? Dio();

  final Dio _dio;

  Map<String, String> get _headers => {'X-User-Id': currentUserScope()};

  Future<List<HabitModel>> listHabits() async {
    final resp = await _dio.get<Map<String, dynamic>>(
      '${ApiConstants.baseUrl}/habits',
      options: Options(headers: _headers),
    );
    final habits = resp.data?['habits'];
    if (habits is! List) return [];
    return habits
        .whereType<Map>()
        .map((m) => HabitModel.fromJson(Map<String, dynamic>.from(m)))
        .toList();
  }

  Future<HabitModel> createHabit(Map<String, dynamic> payload) async {
    final resp = await _dio.post<Map<String, dynamic>>(
      '${ApiConstants.baseUrl}/habits',
      data: payload,
      options: Options(headers: _headers),
    );
    final habit = resp.data?['habit'];
    if (habit is! Map) {
      throw StateError('Invalid create habit response');
    }
    return HabitModel.fromJson(Map<String, dynamic>.from(habit));
  }

  Future<HabitModel> updateHabit(String id, Map<String, dynamic> payload) async {
    final resp = await _dio.put<Map<String, dynamic>>(
      '${ApiConstants.baseUrl}/habits/$id',
      data: payload,
      options: Options(headers: _headers),
    );
    final habit = resp.data?['habit'];
    if (habit is! Map) {
      throw StateError('Invalid update habit response');
    }
    return HabitModel.fromJson(Map<String, dynamic>.from(habit));
  }

  Future<void> deleteHabit(String id) async {
    await _dio.delete(
      '${ApiConstants.baseUrl}/habits/$id',
      options: Options(headers: _headers),
    );
  }

  Future<List<HabitSessionModel>> scheduleWeek({
    required List<EventModel> calendarEvents,
    DateTime? weekStart,
    int lookAheadDays = 7,
  }) async {
    final resp = await _dio.post<Map<String, dynamic>>(
      '${ApiConstants.baseUrl}/habits/schedule',
      data: {
        'calendar_events': _eventsToApi(calendarEvents),
        if (weekStart != null) 'week_start': weekStart.toIso8601String(),
        'look_ahead_days': lookAheadDays,
      },
      options: Options(headers: _headers),
    );
    return _parseSessions(resp.data?['sessions']);
  }

  Future<List<HabitSessionModel>> rescheduleForNewEvent({
    required List<EventModel> calendarEvents,
    required List<HabitSessionModel> currentSessions,
    required EventModel newEvent,
    DateTime? weekStart,
    int lookAheadDays = 7,
  }) async {
    final resp = await _dio.post<Map<String, dynamic>>(
      '${ApiConstants.baseUrl}/habits/reschedule',
      data: {
        'calendar_events': _eventsToApi(calendarEvents),
        'current_sessions': [
          for (final s in currentSessions) s.toRescheduleJson(),
        ],
        'new_event': {
          'id': newEvent.id,
          'title': newEvent.title,
          'start': newEvent.startTime.toIso8601String(),
          'end': newEvent.endTime.toIso8601String(),
          'source': newEvent.source,
        },
        if (weekStart != null) 'week_start': weekStart.toIso8601String(),
        'look_ahead_days': lookAheadDays,
      },
      options: Options(headers: _headers),
    );
    return _parseSessions(resp.data?['sessions']);
  }

  static List<Map<String, dynamic>> _eventsToApi(List<EventModel> events) => [
        for (final e in events)
          {
            'id': e.id,
            'title': e.title,
            'start': e.startTime.toIso8601String(),
            'end': e.endTime.toIso8601String(),
            'source': e.source,
          },
      ];

  static List<HabitSessionModel> _parseSessions(dynamic raw) {
    if (raw is! List) return [];
    return raw
        .whereType<Map>()
        .map((m) => HabitSessionModel.fromJson(Map<String, dynamic>.from(m)))
        .toList();
  }
}

File: backend/app/services/habit_service.py
"""Facade for habit CRUD and scheduling operations."""
from __future__ import annotations

from datetime import datetime, timedelta
from typing import List, Optional, Sequence

from app.models.habit_models import CalendarEvent, Habit, HabitSession
from app.services.habit_repository import HabitRepository
from app.services.habit_rescheduling_service import HabitReschedulingService
from app.services.habit_scheduling_engine import HabitSchedulingEngine


def _monday_of(dt: datetime) -> datetime:
    base = dt.replace(hour=0, minute=0, second=0, microsecond=0)
    return base - timedelta(days=base.weekday())


class HabitService:
    def __init__(
        self,
        repository: HabitRepository | None = None,
        engine: HabitSchedulingEngine | None = None,
        rescheduler: HabitReschedulingService | None = None,
    ) -> None:
        self._repo = repository or HabitRepository()
        self._engine = engine or HabitSchedulingEngine()
        self._rescheduler = rescheduler or HabitReschedulingService(self._engine)

    # ── CRUD ─────────────────────────────────────────────────────────────

    def list_habits(self, user_id: str) -> List[Habit]:
        return self._repo.list_for_user(user_id)

    def get_habit(self, user_id: str, habit_id: str) -> Optional[Habit]:
        return self._repo.get(user_id, habit_id)

    def create_habit(self, user_id: str, payload: dict) -> Habit:
        return self._repo.create(user_id, payload)

    def update_habit(self, user_id: str, habit_id: str, payload: dict) -> Optional[Habit]:
        return self._repo.update(user_id, habit_id, payload)

    def delete_habit(self, user_id: str, habit_id: str) -> bool:
        return self._repo.delete(user_id, habit_id)

    # ── Scheduling ───────────────────────────────────────────────────────

    def schedule_week(
        self,
        user_id: str,
        calendar_events: Sequence[CalendarEvent],
        *,
        week_start: datetime | None = None,
        look_ahead_days: int = 7,
    ) -> dict:
        habits = self._repo.list_for_user(user_id)
        week_start = week_start or _monday_of(datetime.utcnow())
        sessions = self._engine.schedule_habits_fresh(
            habits,
            calendar_events,
            week_start=week_start,
            look_ahead_days=look_ahead_days,
        )
        return {
            "week_start": week_start.isoformat(),
            "sessions": [_session_to_dict(s) for s in sessions],
            "decisions": [
                {"session_id": s.id, "habit_id": s.habit_id, "explanation": s.explanation}
                for s in sessions
            ],
        }

    def reschedule_for_new_event(
        self,
        user_id: str,
        calendar_events: Sequence[CalendarEvent],
        current_sessions: Sequence[HabitSession],
        new_event: CalendarEvent,
        *,
        week_start: datetime | None = None,
        look_ahead_days: int = 7,
    ) -> dict:
        habits = self._repo.list_for_user(user_id)
        week_start = week_start or _monday_of(datetime.utcnow())
        updated, log = self._rescheduler.reschedule_after_new_event(
            habits=habits,
            calendar_events=calendar_events,
            current_sessions=current_sessions,
            new_event=new_event,
            week_start=week_start,
            look_ahead_days=look_ahead_days,
        )
        return {
            "sessions": [_session_to_dict(s) for s in updated],
            "changes": log,
        }


def _session_to_dict(session: HabitSession) -> dict:
    return {
        "id": session.id,
        "habit_id": session.habit_id,
        "habit_title": session.habit_title,
        "start_time": session.start.isoformat(),
        "end_time": session.end.isoformat(),
        "explanation": session.explanation,
        "score": session.score,
        "is_habit": True,
        "is_ai_generated": False,
    }
